### PR TITLE
fix a bug where re-exported struct names via require/typed/provide don't have struct infomation

### LINKED ISSUE #837: Untyped code requiring struct from `require/typed`
### What version of Racket are you using?

`7.3--2019-06-14`

### What program did you run?

```rkt
(module a racket/base
  (struct foo ())
  (provide (struct-out foo)))

(module b typed/racket/base
  (require/typed (submod ".." a) [#:struct foo ()])
  (provide (struct-out foo)))

(module c racket/base
  (require (submod ".." b))
  (provide (struct-out foo)))
```

### What should have happened?

No error.

### If you got an error message, please include it here.

```
struct-out: identifier is not bound to struct type information in: foo
```

### PR DESCRIPTION
**Pull request by capfredf**
closes #837 
supersedes #926

### CODE REVIEW

**Comment by bennn** [comment]
Hi @capfredf , I just ran into this issue again (with @llazarek). 

Do you remember what's the status here? In particular, are there any majors problems that would prevent a fix?

**Comment by capfredf** [comment]
@bennn It has been a while, so my memory is a bit vague. If I recall correctly, there are some issues with imported structs with `#:constructor-name` specified via `require/typed`. The real bigger issue that stopped me from proceeding with this PR was https://github.com/racket/macro-debugger/issues/34, which made debugging the issue much harder.

**Comment by bennn** [comment]
Ok, thanks! I may try to pick this up. (Hopefully one of the tests here illustrates the `#:constructor-name` problem, but if not that's a good hint to start with.)

**Comment by capfredf** [comment]
@bennn Sam and I talked about this pr. Since this pr as-is is an improvement, we've decided to move ahead and try to get this merged into the release of 8.6. The I will work on the fix/support for `extra-constructor-name` later

File: typed-racket-lib/typed-racket/typecheck/def-binding.rkt [hunk 1]
```diff
@@ -109,18 +109,20 @@
      (define sname-is-constructor? (and (or extra-constr-name (free-identifier=? sname constructor-name)) #t))
      (define constr (or extra-constr-name constr^))
      (define type-is-sname? (free-identifier=? tname internal-id))
+
      ;; Here, we recursively handle all of the identifiers referenced
      ;; in this static struct info.
      (define-values (constr-defn constr-export-defn constr-new-id constr-aliases)
        (cond
-         [(not (identifier? constr))
```
Comment (capfredf) [inline typed-racket-lib/typed-racket/typecheck/def-binding.rkt:116]: Revert

**Comment by samth** [comment]
Is this done?

**Comment by capfredf** [comment]
I made a mistake when I rebased the commits. It should be fixed now.

File: typed-racket-lib/typed-racket/base-env/prims-contract.rkt [hunk 2]
```diff
@@ -448,7 +448,7 @@
                       ;The actual identifier bound to the constructor
                       [real-maker (if (syntax-e #'id-is-ctor?) #'internal-maker #'maker-name)]
                       [extra-maker (and (attribute input-maker.extra)
-                                        (not (bound-identifier=? #'make-name #'nm))
+                                        (not (bound-identifier=? #'maker-name #'nm))
```
Comment (samth) [inline typed-racket-lib/typed-racket/base-env/prims-contract.rkt:451]: Is this the only substantive change, other than removing the `type-only` field everywhere?

File: typed-racket-lib/typed-racket/typecheck/def-binding.rkt [hunk 3]
```diff
@@ -107,8 +107,9 @@
              (mk-ignored-quad e))))
 
      (define sname-is-constructor? (and (or extra-constr-name (free-identifier=? sname constructor-name)) #t))
-     (define constr (or extra-constr-name constr^))
+     (define constr (or extra-constr-name constructor-name))
```
Comment (capfredf) [inline typed-racket-lib/typed-racket/typecheck/def-binding.rkt:110]: @samth the other meaningful change is here. The constructor name from the repackaged struct info by `request/typed` are not actually exported as the constructor, so we use the actual constructor name received by `dtsi*`.

Comment (capfredf) [reply typed-racket-lib/typed-racket/typecheck/def-binding.rkt:110]: It is very likely that we can simplify the code for require/typed for structs, but we should do that in this PR

**Pull request merged**